hips/hip-0019.md

```diff
@@ -0,0 +1,65 @@
+---
+hip: "0019"
+title: "Secrets storage driver support for split manifests"
+authors: [ "Louis Cabrol" ]
+created: "2023-02-06"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+Helm, by default in version 3, stores its Release manifest in a Secret object in the target Kubernetes namespace, after a double base64 encoding + GZIP compression.
+
+This object type, [by Kubernetes design](https://kubernetes.io/docs/concepts/configuration/secret/#restriction-data-size) has a maximum size allowed for data restricted to 1Mib.
+
+This imposes a limit on the maximum number of templates a Helm Release can manage, as well as limits the overall complexity a Helm Chart can reach before exceeding this limit.
+
+## Motivation
+
+This has been reported a number of times ([here](https://github.com/helm/helm/issues/10986), [here](https://github.com/helm/helm/issues/8281), and many others within those two links). As Charts keep growing in complexity through modular, optional components, and Releases keep growing in the number of Kubernetes objects they manage, more and more people will keep hitting this limit.
+
+Deploying a Chart that exceeds this preset size currently yields a Kubernetes error, without first warning the user that the Release Manifest size exceeds the maximum, which isn't a good user experience.
+
+My specific motivation for this fix is a complex, modular Helm Chart that deploys Grafana dashboards. As such, it contains (many) dashboard templates as JSON that are templated through Helm into ConfigMap objects. Those template files cannot be `helmignored`, and their cumulated size alone is close to the Secret size limit, before encoding and compression. As such, I'm starting to run into issues when deploying to complex environments that require a lot of those dashboard definitions, which could be solved by splitting the chart among different sub-charts and/or sub-releases, but this would be working around the problem, and not adressing it.
+
+## Specification
+
+The proposal is to create a new "type" of Secret, `helm.sh/partial.v1`, that would enable the Release Manifest to be split across multiple secrets as needed, named after the following format `sh.helm.partial.v1.<release name>.v<release version>-<partial chunk number>`. The presence of further "partial" secrets would then be indicated through an optional `continuedIn` label. Changes should only be made within the Secrets storage driver, so as to retain the current driver interface.
```

Comment:
> The presence of further "partial" secrets would then be indicated through an optional `continuedIn` label

Should Helm self-impose a limit? I didn't see any limits in the sql driver (the release body is stored as a [text](https://www.postgresql.org/docs/current/datatype-character.html) type). But it does seem like Helm should error if the user is doing something most-likely wrong. Like trying to store a several GB release manifest.

Reply:
> But it does seem like Helm should error if the user is doing something most-likely wrong

I'm not sure where that line would be drawn to be honest. Like you said, the SQL driver doesn't impose limits, so it likely would need to be added from the top down and applied to all storage drivers, rather than be built in the split manifest version and only applied there.

Reply:
I agree. This policy should really be at the Helm level, not the individual driver level (generally). Resolving this.

Reply:
(err, for some reason I can't resolve my comment; otherwise I would)